File: internal/driver/mqtt/cluster.go
package mqtt

import (
	"context"
	"fmt"
	"sort"
	"strconv"
	"strings"

	"github.com/amigoer/mq-studio/internal/model"
	"github.com/amigoer/mq-studio/internal/timestamp"
)

// Attributes the cluster pages read. They are a contract with
// frontend/src/mq/mqtt/cluster.ts.
const (
	AttrClientsConnected    = "clientsConnected"
	AttrClientsTotal        = "clientsTotal"
	AttrClientsMaximum      = "clientsMaximum"
	AttrClientsInactive     = "clientsInactive"
	AttrClientsExpired      = "clientsExpired"
	AttrSubscriptions       = "subscriptions"
	AttrSharedSubscriptions = "sharedSubscriptions"
	AttrRetainedCount       = "retainedCount"
	AttrMessagesReceived    = "messagesReceived"
	AttrMessagesSent        = "messagesSent"
	AttrMessagesStored      = "messagesStored"
	AttrMessagesDropped     = "messagesDropped"
	AttrBytesReceived       = "bytesReceived"
	AttrBytesSent           = "bytesSent"
	AttrHeapCurrent         = "heapCurrent"
	AttrUptimeSeconds       = "uptimeSeconds"
	AttrBrokerVersion       = "brokerVersion"

	// AttrSysTopics is the whole tree, as "topic\tvalue" lines. The overview
	// shows it verbatim, so a broker publishing counters this driver does not
	// know by name still has them on screen rather than silently dropped.
	AttrSysTopics = "sysTopics"
)

/*
 * A cluster page for a protocol with no cluster.
 *
 * MQTT says nothing about how a broker is deployed. What can be reported is
 * the one broker this session is talking to, out of the $SYS tree it publishes
 * about itself - so ListNodes returns exactly one node, and says so by
 * reporting a total of one rather than leaving the count unknown.
 *
 * A real cluster's other members are only visible through a vendor API, which
 * is the management tier's job and not this one's.
 */

// ListNodes is the broker this session is connected to.
func (c *Conn) ListNodes(ctx context.Context) ([]*model.Node, error) {
	node, err := c.brokerNode(ctx)
	if err != nil {
		return nil, err
	}
	return []*model.Node{node}, nil
}

// NodeDetail is the same broker. The address is checked rather than ignored so
// a stale link from another page fails visibly instead of silently answering
// about a different node.
func (c *Conn) NodeDetail(ctx context.Context, address string) (*model.Node, error) {
	node, err := c.brokerNode(ctx)
	if err != nil {
		return nil, err
	}
	if address != "" && address != node.Address {
		return nil, fmt.Errorf("this connection has no node at %q", address)
	}
	return node, nil
}

// ClusterOverview is the header figures the overview board reads.
func (c *Conn) ClusterOverview(ctx context.Context) (*model.ClusterOverview, error) {
	tree, err := c.readSys(ctx)
	if err != nil {
		return nil, err
	}

	overview := &model.ClusterOverview{
		Name:        c.brokerAddress(),
		TotalNodes:  1,
		OnlineNodes: 1,
		// A topic is not an object here, so there is no count of them. The
		// retained total is a different figure and is carried as its own
		// attribute rather than passed off as this one.
		Destinations:  model.UnknownMetric,
		Subscriptions: intMetric(tree.number(sysSubscriptions)),
		// No broker reports disk through $SYS, and MQTT stores almost nothing
		// anyway.
		AvgDiskUsage: model.UnknownMetric,
		Attributes:   sysAttributes(tree),
	}
	return overview, nil
}

// brokerNode builds the single node from one $SYS read.
func (c *Conn) brokerNode(ctx context.Context) (*model.Node, error) {
	tree, err := c.readSys(ctx)
	if err != nil {
		return nil, err
	}

	node := &model.Node{
		ID:      1,
		Name:    c.brokerAddress(),
		Address: c.brokerAddress(),
		Version: tree.text(sysVersion),
		Status:  model.NodeOnline,
		// The load averages are per minute; the page shows per second.
		RateIn:  perSecond(tree.number(sysLoadIn1min)),
		RateOut: perSecond(tree.number(sysLoadOut1min)),
		// MQTT brokers hold retained messages and little else, and none of
		// them publish a disk figure.
		DiskUsage:  model.UnknownMetric,
		LastSeen:   timestamp.Now(),
		Attributes: sysAttributes(tree),
	}
	return node, nil
}

// brokerAddress names the endpoint this session is on. There is one, because a
// session is one socket: the profile's other addresses are failover candidates
// rather than nodes of a cluster.
func (c *Conn) brokerAddress() string {
	if len(c.config.Servers) == 0 {
		return ""
	}
	return c.config.Servers[0].Host
}

// sysAttributes maps the tree onto the keys the boards read, and carries the
// whole tree alongside them.
func sysAttributes(tree *sysTree) map[string]string {
	attributes := map[string]string{
		AttrSysTopics: sysTopicLines(tree),
	}
	if version := tree.text(sysVersion); version != "" {
		attributes[AttrBrokerVersion] = version
	}

	counters := map[string]string{
		AttrClientsConnected:    sysClientsConnected,
		AttrClientsTotal:        sysClientsTotal,
		AttrClientsMaximum:      sysClientsMaximum,
		AttrClientsInactive:     sysClientsInactive,
		AttrClientsExpired:      sysClientsExpired,
		AttrSubscriptions:       sysSubscriptions,
		AttrSharedSubscriptions: sysSharedSubscriptions,
		AttrRetainedCount:       sysRetained,
		AttrMessagesReceived:    sysMessagesReceived,
		AttrMessagesSent:        sysMessagesSent,
		AttrMessagesStored:      sysMessagesStored,
		AttrMessagesDropped:     sysMessagesDropped,
		AttrBytesReceived:       sysBytesReceived,
		AttrBytesSent:           sysBytesSent,
		AttrHeapCurrent:         sysHeapCurrent,
	}
	for attribute, key := range counters {
		// A counter the broker does not publish is left out entirely rather
		// than written as -1: the boards read a missing attribute as "not
		// reported", and a -1 on screen is worse than a blank.
		if value := tree.number(key); value != unknown {
			attributes[attribute] = strconv.FormatInt(value, 10)
		}
	}
	if uptime := tree.uptimeSeconds(); uptime != unknown {
		attributes[AttrUptimeSeconds] = strconv.FormatInt(uptime, 10)
	}
	return attributes
}

// sysTopicLines renders the whole tree for the overview's own table, sorted so
// two reads of an unchanged broker look the same.
func sysTopicLines(tree *sysTree) string {
	topics := make([]string, 0, len(tree.Values))
	for topic := range tree.Values {
		topics = append(topics, topic)
	}
	sort.Strings(topics)

	var lines strings.Builder
	for _, topic := range topics {
		lines.WriteString(topic)
		lines.WriteByte('\t')
		lines.WriteString(tree.Values[topic])
		lines.WriteByte('\n')
	}
	return lines.String()
}

// perSecond turns a per-minute load average into the per-second rate the
// canonical node field carries.
func perSecond(perMinute int64) int {
	if perMinute == unknown {
		return model.UnknownMetric
	}
	return int(perMinute / 60)
}

func intMetric(value int64) int {
	if value == unknown {
		return model.UnknownMetric
	}
	return int(value)
}

File: frontend/src/mq/mqtt/cluster.ts
/**
 * MQTT's view of the canonical cluster models.
 *
 * The keys are a contract with internal/driver/mqtt/cluster.go.
 *
 * What is absent matters as much as what is here, and more than it does for
 * any other family. MQTT reports no disk figure, no per-destination depth and
 * no rate that is not derived from a broker's own load average - so the
 * canonical fields carry the unknown sentinel and no board draws a number the
 * protocol cannot produce.
 *
 * Which figures arrive at all depends on the deployment. A plain Mosquitto
 * publishes a full $SYS tree and has no management API; a default EMQX refuses
 * the $SYS subscription and answers everything over HTTP. Reading through
 * `count` below means a board shows what this broker reports and says nothing
 * about the rest, rather than drawing a zero for both cases.
 */
import type { ClusterOverview, Node } from "@bindings/model/models";

const AttrBrokerVersion = "brokerVersion";
const AttrUptimeSeconds = "uptimeSeconds";
const AttrSysTopics = "sysTopics";

const AttrClientsConnected = "clientsConnected";
const AttrClientsTotal = "clientsTotal";
const AttrClientsMaximum = "clientsMaximum";
const AttrSharedSubscriptions = "sharedSubscriptions";
const AttrRetainedCount = "retainedCount";
const AttrMessagesReceived = "messagesReceived";
const AttrMessagesSent = "messagesSent";
const AttrMessagesDropped = "messagesDropped";
const AttrBytesReceived = "bytesReceived";
const AttrBytesSent = "bytesSent";
const AttrHeapCurrent = "heapCurrent";

const AttrNodeRole = "nodeRole";
const AttrNodeEdition = "nodeEdition";
const AttrNodeConnections = "nodeConnections";
const AttrNodeLive = "nodeLiveConnections";
const AttrNodeSessions = "nodeSessions";
const AttrMemoryUsed = "memoryUsed";
const AttrMemoryTotal = "memoryTotal";
const AttrLoad1 = "load1";

type Attributed = { attributes?: Record<string, string | undefined> };

function attr(source: Attributed, key: string): string {
  return source.attributes?.[key] ?? "";
}

/**
 * A counter the broker did not publish reads as unknown, not as zero.
 *
 * The difference is the whole point of the tiers: "this broker does not count
 * dropped messages" and "no messages were dropped" look identical once both
 * are rendered as 0, and only one of them means anything is healthy.
 */
function count(source: Attributed, key: string): number | null {
  const raw = attr(source, key);
  if (raw === "") return null;
  const parsed = Number.parseInt(raw, 10);
  return Number.isNaN(parsed) ? null : parsed;
}

/** The overview's KPI row, straight from whichever tier answered. */
export interface MqttBrokerStats {
  version: string;
  uptimeSeconds: number | null;
  clientsConnected: number | null;
  clientsTotal: number | null;
  clientsMaximum: number | null;
  subscriptions: number | null;
  sharedSubscriptions: number | null;
  retained: number | null;
  messagesReceived: number | null;
  messagesSent: number | null;
  messagesDropped: number | null;
  bytesReceived: number | null;
  bytesSent: number | null;
  heapBytes: number | null;
  /** Topics, which only a management API can count. */
  topics: number | null;
}

export function brokerStats(overview: ClusterOverview): MqttBrokerStats {
  return {
    version: attr(overview, AttrBrokerVersion),
    uptimeSeconds: count(overview, AttrUptimeSeconds),
    clientsConnected: count(overview, AttrClientsConnected),
    clientsTotal: count(overview, AttrClientsTotal),
    clientsMaximum: count(overview, AttrClientsMaximum),
    // The canonical field is filled by both tiers, so it is read rather than
    // the attribute - and -1 there is the driver's not-reported marker.
    subscriptions: overview.subscriptions >= 0 ? overview.subscriptions : null,
    sharedSubscriptions: count(overview, AttrSharedSubscriptions),
    retained: count(overview, AttrRetainedCount),
    messagesReceived: count(overview, AttrMessagesReceived),
    messagesSent: count(overview, AttrMessagesSent),
    messagesDropped: count(overview, AttrMessagesDropped),
    bytesReceived: count(overview, AttrBytesReceived),
    bytesSent: count(overview, AttrBytesSent),
    heapBytes: count(overview, AttrHeapCurrent),
    topics: overview.destinations >= 0 ? overview.destinations : null,
  };
}

/**
 * The broker's own $SYS tree, as it published it.
 *
 * Shown verbatim rather than curated, because a broker publishes counters this
 * app has never heard of and dropping them would make the page less useful
 * than `mosquitto_sub -t '$SYS/#'`.
 */
export function sysTopics(overview: ClusterOverview): { topic: string; value: string }[] {
  const raw = attr(overview, AttrSysTopics);
  if (raw === "") return [];
  return raw
    .split("\n")
    .filter((line) => line !== "")
    .map((line) => {
      const tab = line.indexOf("\t");
      return tab < 0
        ? { topic: line, value: "" }
        : { topic: line.slice(0, tab), value: line.slice(tab + 1) };
    });
}

/** One node's detail, where the broker names more than one. */
export interface MqttNodeDetail {
  role: string;
  edition: string;
  connections: number | null;
  liveConnections: number | null;
  sessions: number | null;
  memoryUsed: string;
  memoryTotal: string;
  load1: string;
  uptimeSeconds: number | null;
}

export function nodeDetail(node: Node): MqttNodeDetail {
  return {
    role: attr(node, AttrNodeRole),
    edition: attr(node, AttrNodeEdition),
    connections: count(node, AttrNodeConnections),
    liveConnections: count(node, AttrNodeLive),
    sessions: count(node, AttrNodeSessions),
    memoryUsed: attr(node, AttrMemoryUsed),
    memoryTotal: attr(node, AttrMemoryTotal),
    load1: attr(node, AttrLoad1),
    uptimeSeconds: count(node, AttrUptimeSeconds),
  };
}

/**
 * Uptime as a person reads it.
 *
 * Seconds rather than the milliseconds RabbitMQ's node page formats, because
 * that is the unit both MQTT tiers report in: Mosquitto publishes "3600
 * seconds" under $SYS and the driver converts EMQX's milliseconds on the way
 * out, so there is one unit to render here rather than two.
 */
export function formatUptimeSeconds(seconds: number | null): string {
  if (seconds == null || seconds < 0) return "—";
  const days = Math.floor(seconds / 86400);
  const hours = Math.floor((seconds % 86400) / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  if (days > 0) return `${days}d ${hours}h`;
  if (hours > 0) return `${hours}h ${minutes}m`;
  if (minutes > 0) return `${minutes}m`;
  return `${seconds}s`;
}
